### ISSUE TITLE
Need option to bind server to 0.0.0.0 in development mode when using Vagrant

### ISSUE THREAD

**Issue by kiere** [rails/webpacker#709]
Original Issue Title: **manifest.json not respecting "public" option to devServer settings (Vagrant nginx proxy)**

I am using webpacker 2.0 on Rails 4.2.  I have run `webpack:install` and `webpack:install:elm`.  Our setup uses Vagrant with the VM aliased to something similar to `myapp.test` using `vagrant-hostmanager` plugin.  We want to be able to use `foreman start` to spin up Unicorn and `webpack-dev-server` inside the VM and be able to access the Rails app from our local dev laptop (outside the VM) using `http://myapp.test`.  This all works fine except for the "packs" assets because accessing `http://0.0.0.0:8080` from the local laptop will not get through to the VM.  So, adding the `public` option to the devServer settings (through webpacker.yml, or development.js directly, or as a `--public` option to `bin/webpack-dev-server` WILL allow me access the "packs" resources from the dev laptop using the host alias (http://myapp.test:8080/packs/application.js), but the scripts inside the Rails view template will not reflect the `public` option and continues to render the `host` and `port` option.

```HTML
<script src="http://0.0.0.0:8080/packs/application.js"></script>
<script src="http://0.0.0.0:8080/packs/hello_elm.js"></script>
```

```Yaml
dev_server:
    host: 0.0.0.0
    port: 8080
    https: false
    public: myapp.test:8082  # <--- Line added
```

```JavaScript
devServer: {
    clientLogLevel: 'none',
    https: settings.dev_server.https,
    host: settings.dev_server.host,
    port: settings.dev_server.port,
    contentBase: output.path,
    public: settings.dev_server.public, // <--- Line added
    publicPath: output.publicPath,
    compress: true,
    headers: { 'Access-Control-Allow-Origin': '*' },
    historyApiFallback: true,
    watchOptions: {
      ignored: /node_modules/
    }
  }
```

For example, if I open up 3 browser windows... 1.) http://myapp.test  2.) http://myapp.test:8080/packs/application.js 3.) http://myapp.test/packs/manifest.json, I will see my app in the first window, the output of application.js in the second window, but in the 3rd window, I will see the JSON using http://0.0.0.0:8080 (instead of http://myapp.test:8080).  This will result in 2 console errors in the first tab (where my app is) referencing connection errors to http://0.0.0.0:8080.

I've very confused why setting a new key in `webpacker.yml` for `public` and then updating `config/webpack/development.js` to use this new `public` key will update `webpack-dev-server` to be available on my VM (from my local dev machine) but it will not render the manifest.json correctly so that the inline pack tags use the `public` address of the `webpack-dev-server`.

Any ideas?  How can I properly modify the config so that the `javascript_pack_tag` and `stylesheet_pack_tag` use the correct domain name of the `webpack-dev-server`.  I feel like I have tried so many options and the only thing that works is to port forward port 8080 to the local VM, but that is a brittle solution if there are port conflict errors (and the inevitable remapping) when booting the Vagrant VM.

I should note that I have nginx running inside the VM (sitting in front of Unicorn) so I added the nginx config reference here:
https://webpack.js.org/guides/development-vagrant/#advanced-usage-with-nginx
So that I can access the `webpack-dev-server` using a different port than my app, but also behind nginx as the proxy.  So, in reality, in my app, I have the Rails app via Unicorn accessible on port 80 behind nginx and the `webpack-dev-server` accessible on port 8082 (which proxies to http://127.0.0.1:8080).  Therefore, on my local dev machine, I can access my app at http://myapp.test and the webpack-dev-server at http://myapp.test:8082.  But I just can't get Rails to reflect this correctly when it renders the javascript and stylesheet pack tags.

Articles and docs that I have referenced:
* https://github.com/rails/webpacker/blob/master/README.md
* https://webpack.js.org/configuration/dev-server/#devserver-public
* https://webpack.js.org/guides/development-vagrant/
* https://webpack.js.org/guides/development-vagrant/#advanced-usage-with-nginx

**Comment by gauravtiwari** [comment]
@kiere Please upgrade to master and re-install webpacker - we have fixed this issue but yet to be released. 

```rb
gem 'webpacker', github: 'rails/master'
```

```
bundle update webpacker
bundle exec rails webpacker:install
```

**Comment by kiere** [comment]
@gauravtiwari Thanks!  That took care of it.  Just a few notes for those that follow.  I think you have a typo above:

```
gem 'webpacker', github: 'rails/webpacker'
```
Note: use `rails/webpacker` vs. `rails/master`

Also, once I completed the steps above, I had to rsync the changes to the VM and then, inside the VM, run:

```
cd /vagrant (or whatever your project root is)
bundle install
yarn install
```
Then I could restart Unicorn and `webpack-dev-server`.

It looks like the nginx proxy and other items are no longer needed.

**Comment by gauravtiwari** [comment]
@kiere Fixed the typo, thanks. That's right - the proxy is built-in now 👍

**Comment by kiere** [comment]
@gauravtiwari Is the following console error expected given my configuration with Vagrant?  Or is there something more I should do to clear this?

```
abstract-xhr.js?c769:132 GET http://localhost:3035/sockjs-node/info?t=1504043676759 net::ERR_CONNECTION_REFUSED
```

**Comment by gauravtiwari** [comment]
@kiere Are you running webpack dev server?

**Comment by kiere** [comment]
@gauravtiwari Yes, we are running that inside the VM using `foreman start` (which spins up Unicorn, a delayed job worker and webpack-dev-server).

**Comment by gauravtiwari** [comment]
@kiere Can you open this link in a separate tab `http://localhost:3035/sockjs-node/info?t=1504043676759` please and see what that gives you?

**Comment by gauravtiwari** [comment]
I am assuming it could be something to do with network i.e. localhost usage inside VM which might be conflicting.

**Comment by kiere** [comment]
@gauravtiwari No, I cannot open that link on my local dev machine because it doesn't have port forward access to the VM.  That's why I filed the ticket originally.  I need to be able to access the webpack-dev-server via it's `public` option.  So, I need to be able to pass `--public myapp.test:3035` to `bin/webpack-dev-server` (either via command line option in the `Procfile` or via the files in `config/webpack`) and then have webpacker pickup that option so the inline scripts use that `public` option when it generates the URL in question:

```
http://localhost:3035/sockjs-node/info?t=1504043676759
```

becomes:
```
http://myapp.test:3035/sockjs-node/info?t=1504043676759
```

I haven't looked at the current version of webpacker in `master` to see what you changed to make this work all of a sudden (vs. using v2.0.0), but it seems like there's something missing.  I am going to try playing around with variations of the `public` option, but when I start webpack-dev-server via `bin/webpack-dev-server`, it no longer seems to report "Project is running at http://myapp.test:3035", which is what it used to report when I would set the public option.

**Comment by gauravtiwari** [comment]
@kiere Please update host in `config/webpacker.yml`

**Comment by kiere** [comment]
@gauravtiwari Ok, I took a look at `bin/webpack-dev-server`.  Just updating the `host` key in `config/webpacker.yml` isn't enough to get this to work because even though that _did_ update the URLs in the inline script, the connectivity error was still being logged in the console.

I first attempted to fix this by adding `disableHostCheck: true` to the `config/webpack/development.js` file.  But that didn't work.  Then I removed that and added `allowed_hosts: [ 'myapp.test' ]`, but that didn't work either.

After inspecting `bin/webpack-dev-server`, I was able to hack this into working by altering this line:

From:
```
LISTEN_HOST_ADDR = NODE_ENV == 'development' ? 'localhost' : '0.0.0.0'
```

To:
```
LISTEN_HOST_ADDR = NODE_ENV == 'development' ? '0.0.0.0' : '0.0.0.0'
```

So, obviously this isn't ideal (or good code to leave).  But, that makes sense.  Since the server was listening on `localhost` when `NODE_ENV == 'development'`, then I wasn't able to access it from outside the VM.  Once I updated the listen address to be `0.0.0.0`, I was then able to access the server via `http://myapp.test:3035` and everything started working properly.

For anyone else that follows, I did need to set `poll: true` in the `watchOptions` object of `config/webpack/development.js` due to the article listed in my original post above...

https://webpack.js.org/guides/development-vagrant/#running-the-server

_See the note on `--watch-poll`_

**Comment by gauravtiwari** [comment]
@kiere Oh yes, I realised just after writing that comment. The current server interface by default listens on `localhost`. What you could do is this in your Procfile: 

```bash
./bin/webpack-dev-server --host 0.0.0.0 
```

without hardcoding inside binstub and that way it will available on all interfaces - `localhost` `VM` and `LAN IP`.

**Comment by gauravtiwari** [comment]
And you can set the poll option in the Procfile instead of development.js

```
./bin/webpack-dev-server --host 0.0.0.0 --public some-vm-ip:3035 --watch-poll
```

P.S. we just released 3.0.0 yesterday: 

```bash
bundle update webpacker
bundle exec rails webpacker:install # (one last time to update configs) 
```

**Comment by kiere** [comment]
@gauravtiwari Passing `--host` on the command line would not work in this case because that is already being used for the external domain name of `myapp.test`.  If I set `--host` to `0.0.0.0` then webpack-dev-server will no longer be accessible from outside the VM because `0.0.0.0` would be referring to my local dev laptop, not the VM.

If you look at `bin/webpack-dev-server`, you will see that `LISTEN_HOST_ADDR` is hardcoded to one of two values in a ternary expression.  There is no configuration to adapt this setting for people running `webpacker` inside a Vagrant VM (or any VM) without port forwarding.

Their only option will be to edit that line in `bin/webpack-dev-server` and fix it any time that file gets replaced by `webpacker`.

Also, thanks for the heads-up on 3.0.0.  And great work on all this.  I love it already.

**Comment by gauravtiwari** [comment]
@kiere forgive me, seems like we changed a few things lately. It wasn't like that before, will push something to make this configurable in that case 👍 

Great, thanks 😄

**Comment by gauravtiwari** [comment]
@kiere  Just released 3.0.1 with a new option

**Comment by DimitriOstapenko** [comment]
I add  
bind tcp://0.0.0.0:3000  
to config/puma.rb
